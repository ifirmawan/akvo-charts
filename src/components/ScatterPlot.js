import React from 'react';
import { useECharts } from '../hooks';
import styles from '../styles.module.css';

const scatterTransform = (input) => {
  if (Array.isArray(input)) {
    // Check if it's a 2D array
    if (Array.isArray(input[0])) {
      return input.map(([label, x, y]) => [x, y, label]);
    }

    // Check if it's an array of key-value objects
    if (typeof input[0] === 'object' && !Array.isArray(input[0])) {
      return input.map(({ label, x, y }) => [x, y, label]);
    }
  }

  // Check if it's an array containing an object
  if (typeof input === 'object') {
    return Object.keys(input).map((key) => [...input[key], key]);
  }

  throw new Error('Invalid input format');
};

const getOptions = ({
  data,
  symbolSize,
  showLabel,
  transformedConfig,
  overrideItemStyle
}) => {
  return {
    series: [
      {
        type: 'scatter',
        symbolSize,
        emphasis: {
          focus: 'self'
        },
        label: {
          show: showLabel,
          formatter: (p) => p.data[2],
          minMargin: 10,
          position: 'top'
        },
        ...overrideItemStyle
      }
    ],
    xAxis: {
      ...transformedConfig.xAxis,
      splitLine: {
        show: true
      }
    },
    dataset: {
      source: scatterTransform(data)
    }
  };
};

const ScatterPlot = ({ config, data, symbolSize = 10, showLabel = true }) => {
  const chartRef = useECharts({
    config,
    getOptions: ({ transformedConfig, overrideItemStyle }) =>
      getOptions({
        data,
        symbolSize,
        showLabel,
        transformedConfig,
        overrideItemStyle
      })
  });

  return (
    <div
      ref={chartRef}
      role="figure"
      className={styles.container}
    />
  );
};

export default ScatterPlot;
