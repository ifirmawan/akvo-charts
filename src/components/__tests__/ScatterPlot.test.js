import React from 'react';
import { render, screen } from '@testing-library/react';
import ScatterPlot from '../ScatterPlot';
import renderer from 'react-test-renderer';

describe('ScatterPlot chart', () => {
  test('renders ScatterPlot component with 2d array data format', () => {
    const input2D = [
      ['A', 1, 2],
      ['B', 1, 3],
      ['C', 5, 7]
    ];

    const config = {
      title: 'ScatterPlot Chart Example'
    };

    render(
      <ScatterPlot
        config={config}
        data={input2D}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders ScatterPlot component with row based key-value format (object array)', () => {
    const inputKeyValue = [
      { label: 'A', x: 1, y: 2 },
      { label: 'B', x: 1, y: 3 },
      { label: 'C', x: 5, y: 7 }
    ];
    const config = {
      title: 'ScatterPlot Chart Example'
    };

    render(
      <ScatterPlot
        config={config}
        data={inputKeyValue}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders ScatterPlot component with column based key-value format', () => {
    const inputObj = {
      A: [1, 2],
      B: [1, 3],
      C: [5, 7]
    };

    const config = {
      title: 'ScatterPlot Chart Example'
    };

    render(
      <ScatterPlot
        config={config}
        data={inputObj}
        symbolSize={50}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('matches ScatterPlot snapshot', () => {
    const data = [
      ['A', 1, 2],
      ['B', 1, 3],
      ['C', 5, 7]
    ];

    const config = {
      title: 'ScatterPlot Chart Example'
    };

    const tree = renderer
      .create(
        <ScatterPlot
          config={config}
          data={data}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
