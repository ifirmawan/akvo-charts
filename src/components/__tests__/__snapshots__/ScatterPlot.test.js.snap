// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterPlot chart matches ScatterPlot snapshot 1`] = `
<div
  className="container"
  role="figure"
/>
`;
